S3Source.cc

```diff
@@ -0,0 +1,248 @@
+#include <iostream>
+#include "S3Source.h"
+#include "SourceFactory.h"
+#include "Deserializer.h"
+#include "UnrolledDeserializer.h"
+
+using namespace cce::tf;
+
+S3Source::S3Source(unsigned int iNLanes, std::string iObjPrefix, int iVerbose, unsigned long long iNEvents, S3ConnectionRef conn):
+  SharedSourceBase(iNEvents),
+  objPrefix_(iObjPrefix),
+  verbose_(iVerbose),
+  conn_(conn),
+  readTime_{std::chrono::microseconds::zero()}
+{
+  auto start = std::chrono::high_resolution_clock::now();
+
+  conn->get(objPrefix_ + "index", [this](S3Request* req) mutable {
+      if ( req->status == S3Request::Status::ok ) {
+        if ( not index_.ParseFromString(req->buffer) ) {
+          throw std::runtime_error("Could not deserialize index in S3Source construction");
+        }
+      }
+      else { throw std::runtime_error("Could not retrieve index in S3Source construction"); }
+    });
+  if ( verbose_ >= 3 ) {
+    std::cout << index_.DebugString() << "\n";
+  }
+
+  if ( index_.totalevents() < iNEvents ) {
+    std::cout << "WARNING: less events in source than requested: "
+      << index_.totalevents() << " vs. " << iNEvents << ". Will read all available events instead.\n";
+  }
+
+  currentProductStripes_.resize(index_.products_size());
+ 
+  laneInfos_.reserve(iNLanes);
+  for(unsigned int i = 0; i< iNLanes; ++i) {
+    DeserializeStrategy strategy;
+    switch(index_.serializestrategy()) {
+      case objstripe::SerializeStrategy::kRoot:
+        strategy = DeserializeStrategy::make<DeserializeProxy<Deserializer>>();
+        break;
+      case objstripe::SerializeStrategy::kRootUnrolled:
+        strategy = DeserializeStrategy::make<DeserializeProxy<UnrolledDeserializer>>();
+        break;
+    }
+    laneInfos_.emplace_back(index_, std::move(strategy));
+  }
+
+  readTime_ += std::chrono::duration_cast<decltype(readTime_)>(std::chrono::high_resolution_clock::now() - start);
+}
+
+S3Source::LaneInfo::LaneInfo(objstripe::ObjectStripeIndex const& index, DeserializeStrategy deserialize):
+  deserializers_{std::move(deserialize)}
+{
+  dataProducts_.reserve(index.products_size());
+  dataBuffers_.resize(index.products_size(), nullptr);
+  deserializers_.reserve(index.products_size());
+  size_t i{0};
+  for(auto const& pi : index.products()) {
+    TClass* cls = TClass::GetClass(pi.producttype().c_str());
+    if ( cls == nullptr ) {
+      throw std::runtime_error("No TClass reflection available for " + pi.productname());
+    }
+    dataBuffers_[i] = cls->New();
+    dataProducts_.emplace_back(i, &dataBuffers_[i], pi.productname(), cls, &delayedRetriever_);
+    deserializers_.emplace_back(cls);
+    ++i;
+  }
+}
+
+S3Source::LaneInfo::~LaneInfo() {
+  auto it = dataProducts_.begin();
+  for( void * b: dataBuffers_) {
+    it->classType()->Destructor(b);
+    ++it;
+  }
+}
+
+std::pair<const char*, size_t> DelayedProductStripeRetriever::bufferAt(size_t globalEventIndex) const {
+  std::call_once(flag_, [this](){
```

Comment:
This could stall many threads at the same time if they are all waiting for the same product buffer, probably something better should be done, but I'm not sure if this is something I could do without having the async S3Connection.

Reply:
OK, so it seems you want to
- read the data associated to a stripe the first time the data is requested by any event using that stripe.
- once read, obtain the part of the data that is associated with that specific event.

We'd want to do this in a way that doesn't lead to blocking any of the threads.

This could be done by using an `std::atomic<char>` and a `WaitingTaskHolder` for each DelayedProductStripeRetriever. 
- Setup an enum with three states, kUnretrieved, kRetrieving, and kRetrieved.
- in getAsync test the `std::atomic<char>` to see if it is in the kRetrieved state, if it is then just read the data for that event
- If the atomic is not in that state, use `compare_exchange_strong to try to change for `kUnretrieved` to `kRetrieving` if succeed call the S3 routine to read the data, then read the data for that event and set the atomic to kRetrieved and tell the WaitingTaskHolder to stop waiting
- if the atomic failed to change state, push a task into the WaitingTaskHolder where that task will just read the data for that event.